### PR TITLE
Add seeding and checkpoints

### PR DESCRIPTION
**Pull request by soham-chitnis10**
This PR resolves issue #60

### CODE REVIEW

**Comment by gchhablani** [comment]
@soham-chitnis10 I am also adding a seed function in #42. Can you please check if I can remove it from my PR?

**Comment by soham-chitnis10** [comment]
@gchhablani I can add deterministic and benchmark

**Comment by someshsingh22** [comment]
@soham-chitnis10 several changes here.
1. Seeding for CUDA isnt proper here, it will throw errors if cuda isnt found make separate functions for that
2. I am adding tests to your PR
3. In the latest Zoo attack PR there is a utils.py remove that and change the code to keep the functionality intact

File: code_soup/common/utils/checkpoints.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import torch
+
+
+class Checkpoints:
+    @classmethod
+    def save(self, PATH, model):
+        torch.save(model.state_dict(), PATH)
+
```
Comment (someshsingh22) [inline code_soup/common/utils/checkpoints.py:8]: This implementation is model specific, we may need to save the optimiser too, Checkpoints save both the model state and the checkpoint state

File: code_soup/common/utils/seeding.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+class Seeding:
+    seed = 42
+
+    @classmethod
+    def set_seeding(self, seed):
+        self.seed = seed
+        np.random.seed(self.seed)
+        torch.manual_seed(self.seed)
+        torch.cuda.manual_seed(self.seed)
```
Comment (someshsingh22) [inline code_soup/common/utils/seeding.py:15]: cuda should be seeded only if cuda is available on CPUs this will fail

File: tests/test_common/test_utils.py [hunk 3]
```diff
@@ -2,13 +2,13 @@
 
 import numpy as np
 
-from code_soup.common.utils import torch_seed
+from code_soup.common.utils.seeding import Seeding
```
Comment (someshsingh22) [inline tests/test_common/test_utils.py:7]: Remove the previous test_utils.py and add util for checkpoints too

**Review by someshsingh22 — CHANGES_REQUESTED** [review]

File: tests/test_common/test_utils/test_checkpoints.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+from code_soup.common.utils.checkpoints import Checkpoints
+
+
+class TestCheckpoints(unittest.TestCase):
+    def test_save(self):
+        """
+        Test that the model is saved
+        """
+        model_save = models.resnet18(pretrained=True)
```
Comment (someshsingh22) [inline tests/test_common/test_utils/test_checkpoints.py:17]: choose a smaller model like discriminator from models.gan, this makes the process a lot slower, and save the files in ./input because that folder is ignored during tests and the files are not committed.

File: tests/test_common/test_utils/test_seeding.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import random
+import unittest
+
+import numpy as np
+import torch
+import torchvision.transforms as transforms
```
Comment (someshsingh22) [inline tests/test_common/test_utils/test_seeding.py:6]: Line no.6 and 7 are irrelevant

File: tests/test_common/test_utils/test_seeding.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+from code_soup.common.utils.seeding import Seeding
+
+
+class TestSeeding(unittest.TestCase):
+    """Test the seed function."""
+
+    def test_seed(self):
+        """Test that the seed is set."""
+        Seeding.seed(42)
```
Comment (someshsingh22) [inline tests/test_common/test_utils/test_seeding.py:17]: Python's random library should also be tested

File: tests/test_common/test_utils/test_checkpoints.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
+import os
+import unittest
+
+import numpy as np
+import torch
+import torch.optim as optim
+import torchvision.models as models
+
+from code_soup.common.utils.checkpoints import Checkpoints
```
Comment (someshsingh22) [inline tests/test_common/test_utils/test_checkpoints.py:9]: This line should be from code_soup.common.utils import Checkpoints
and
from code_soup.common.utils.checkpoints import Checkpoints should be there in code_soup/common/utils/__init__.py

File: tests/test_common/test_utils/test_seeding.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import random
+import unittest
+
+import numpy as np
+import torch
+import torchvision.transforms as transforms
+from torchvision.transforms.transforms import ToTensor
+
+from code_soup.common.utils.seeding import Seeding
```
Comment (someshsingh22) [inline tests/test_common/test_utils/test_seeding.py:9]: This line should be from code_soup.common.utils import Seeding
and
from code_soup.common.utils.seeding import Seeding should be there in code_soup/common/utils/__init__.py

**Review by someshsingh22 — CHANGES_REQUESTED** [review]

**Review by someshsingh22 — APPROVED** [review]
